### ISSUE TITLE
Stale code (that has no effect)

### ISSUE THREAD

**Issue by tcharding** [bitcoin/bitcoin#31558]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behaviour

I am not a C++ programmer. This code looks like it has 4 lines that do nothing.

In commit: 0fbaef9676a a call to `std::ceil` was added which makes the following `if` statement do nothing. There is no harm done but it makes the code confusing for non-c++ devs to read.

```c++
CAmount CFeeRate::GetFee(uint32_t num_bytes) const
{
    const int64_t nSize{num_bytes};

    // Be explicit that we're converting from a double to int64_t (CAmount) here.
    // We've previously had issues with the silent double->int64_t conversion.
    CAmount nFee{static_cast<CAmount>(std::ceil(nSatoshisPerK * nSize / 1000.0))};

    if (nFee == 0 && nSize != 0) {
        if (nSatoshisPerK > 0) nFee = CAmount(1);
        if (nSatoshisPerK < 0) nFee = CAmount(-1);
    }

    return nFee;
}
```

### Expected behaviour

There is no behaviour issue with the current code.

**Comment by tcharding** [comment]
cc @achow101

**Comment by jaeheonshim** [comment]
With the current unit tests, I don't think ceil away from zero will pass unit tests. The way the logic is currently implemented, negative values are rounded up towards zero unless they are greater than -1, in which case they are instead rounded down towards -1. I'm not sure if this is the intended behavior, but if a fix like this were to be implemented:

```
double value = nSatoshisPerK * nSize / 1000.0;
nFee = static_cast<CAmount>(std::copysign(std::ceil(std::abs(value)), value));
```

This test case will fail:

```
feeRate = CFeeRate(-123);
BOOST_CHECK_EQUAL(feeRate.GetFee(9), CAmount(-1));
```

As -123 * 9 / 1000.0 = -1.107, and the old code would round to -1, whereas the new code would round to -2.

**Comment by jaeheonshim** [comment]
@maflcko I'd like to work on this issue, but this being my first time contributing to Bitcoin Core, I'm a bit lost. Where should I look to figure out exactly how the rounding for the fee mechanism should be implemented?

**Comment by yancyribbens** [comment]
Maybe this issue should be closed and a new one started.  The original issue states the code section isn't used, but it pretty clearly is, especially since there's even a test that shows it is.  It is a bit of a mystery what the authors intent was and a code comment might have been nice to explain why if the `ceil` function resulted in zero and the size is not zero, then return -1 if fee_rate is negative or 1 if fee_rate is positive.

**Comment by yancyribbens** [comment]
Actually, I think the "stale code" section is to prevent a fee of zero, and if the fee would be zero, then make it either 1 sat or -1 sat depending of if the fee_rate is positive or negative.

**Comment by jaeheonshim** [comment]
Right, but I think the question is whether negative fee rates should be rounded away from or towards zero. Also, I'm not sure what kind of scenario a negative fee rate would be used.

**Comment by tcharding** [comment]
> In commit: https://github.com/bitcoin/bitcoin/commit/0fbaef9676a1dcb84bcf95afd8d994831ab327b6 a call to std::ceil was added which makes the following if statement do nothing.

Oh it seems I can't do integer division. `nFee` will be zero any time `0 < (nSatoshisPerK * nSize) < 1000` 

I have no opinion on the rounding issue and will leave that to others to raise an issue if needed. Excuse the noise.

**Comment by yancyribbens** [comment]
> Right, but I think the question is whether negative fee rates should be rounded away from or towards zero

`ceil` rounds towards zero: `ceil(-2.4) = -2.000000`.  Rounding toward zero when negative is correct imo.  Rounding up is done to make sure the fee is adequate instead of falling short.

> Also, I'm not sure what kind of scenario a negative fee rate would be used.

Good question.  frameworks like rust-bitcoin use an unsigned int for FeeRate so it can't be negative.

**Comment by tcharding** [comment]
If I'm not mistaken (again) this line of code can never be hit.

`if (nSatoshisPerK > 0) nFee = CAmount(1);`

**Comment by yancyribbens** [comment]
> If I'm not mistaken (again) this line of code can never be hit.

If the goal is to prevent a fee of zero, it probably should be: `if (nSatoshisPerK == 0) nFee = CAmount(1);`
